### ISSUE TITLE
[Impeller] Provide an API for loading FragmentPrograms at runtime (and tooling support for producing them).

### ISSUE THREAD

**Issue by jezell** [flutter/flutter#124759]
## Use case

We would like to be able to dynamically load shaders at runtime like https://www.shadertoy.com or https://www.alma.sh/. Currently FragmentProgram only has a static method for loading assets, which seems a bit limiting as it is just shader bytecode that is being put into the bundle. For instance, web implements this as such:

```dart
 @override
  Future<ui.FragmentProgram> createFragmentProgram(String assetKey) {
    if (_programs.containsKey(assetKey)) {
      return _programs[assetKey]!;
    }
    return _programs[assetKey] = assetManager.load(assetKey).then((ByteData data) {
      return CkFragmentProgram.fromBytes(assetKey, data.buffer.asUint8List());
    });
  }
 ```
  
## Proposal

The functionality already exists behind the scenes, so I just suggest it becomes part of the top level FragmentProgram api:

For instance this already exists:
```dart
static Future<CkFragmentProgram> fromBytes(String name, Uint8List data)
```

**Comment by jonahwilliams** [comment]
This only exists in the Skia backend, because it ships a shader compiler. We'd like to remove this shader compiler eventually with Impeller to reduce code size. We do not plan to add features that require us to retain a shader compiler indefinitely.

**Comment by jezell** [comment]
@jonahwilliams I think you misunderstand the request. I am requesting a method not for compiling the C code for the shader. I am requesting a method for loading either a SPIRV shader (which can definitely be converted at runtime both in web or impeller), or the same platform specific binary data that the compilers are outputting. The request is for runtime loading of shaders not runtime compiling of shaders.

**Comment by jonahwilliams** [comment]
SPIRV would only be supported on a Vulkan backend though. On a metal backend you need to load a metal shader, and on an openGL backend you need to load GLSL. The only way to support consuming arbitrary shaders at runtime is package a runtime compiler. 

We _could_ support an API that simply accepted the binary format that Flutter expects, which today is a blob containing shader code for a particular platform along with uniform metadata.

**Comment by jezell** [comment]
@jonahwilliams yeah that's what I'm requesting, the ability to load them at runtime. Compilation is out of scope for this request. It's not too hard to run SPIRV-cross somewhere and get the output for the target platform, just need some kind of hook to be able to load a runtime asset and a definition somewhere of what the format should be for each platform.

**Comment by jonahwilliams** [comment]
impellerc already produces the format the engine needs, but its not easily usable and the format is currently unstable. We could expose this functionality with caveats that future versions of Flutter may break the feature, or we could stabilize the format. There are likely more caveats

**Comment by chipweinberger** [comment]
**+1** I would also like this feature!

Our app renders midi music & we want want to support custom "themes" uploaded at runtime i.e. fully custom shaders

**Comment by chipweinberger** [comment]
For anyone else that encounters this problem, a workaround is to use a platform webview & webgl. 

Not ideal, but it works for my use case.

**Comment by jezell** [comment]
@chipweinberger maybe if you just need to render a shader, but so good if you need it to be able to interact with the rest of the canvas.

**Comment by Pingear** [comment]
> impellerc already produces the format the engine needs, but its not easily usable and the format is currently unstable. We could expose this functionality with caveats that future versions of Flutter may break the feature, or we could stabilize the format. There are likely more caveats

Hi @jonahwilliams  When exactly are you going to do this? This is exactly what I need now, even if the format may change in the future, we can already begin to develop this direction.

**Comment by jonahwilliams** [comment]
I have no plans to work on this in the near future (hence it is unassigned and low priority). This isn't something that we can reasonably work on until we're closer to done with the Vulkan backend for Android.

**Comment by Pingear** [comment]
> I have no plans to work on this in the near future (hence it is unassigned and low priority). This isn't something that we can reasonably work on until we're closer to done with the Vulkan backend for Android.

It's a pity :(
@jonahwilliams  Can you tell us about this task in a little more detail, Please, in case someone would like to do this? Can you tell where to look, what is ahead, and where to start? I understand that the question is endlessly noob, but maybe in the community there will be a person with suitable skills or someone will want to acquire such skills ;)

**Comment by jonahwilliams** [comment]
I can't really scope out the work for the problem without spending a substantial amount of time on it .... which would mean that I'm not working on higher priority projects

**Comment by junaid1460** [comment]
>Can you tell us about this task in a little more detail, Please, in case someone would like to do this? Can you tell where to look, what is ahead, and where to start? I understand that the question is endlessly noob, but maybe in the community there will be a person with suitable skills or someone will want to acquire such skills ;)

Don't think it's gon be that easy, it's more like creating a new language and a compiler / transpiler. But yeah could create one. 

What would be best is if flutter exposes support to load Vulkan, OpenGL and metal shaders separately. Then community can pick it up. @jonahwilliams is it possible ?

**Comment by kekland** [comment]
> Is code size the main reason to not include the compiler? If so, we could just have a flag in pubspec.yaml to include it. or similar. - runtime-shader-compilation: true

100% agree with this. Currently building a map renderer on top of `flutter_gpu`, and we have a need to compile shaders in run-time to support something similar to data-driven styling from map renderers like MapLibre. My current workaround is to have a compile-time script that generates shader code based on the provided style specification and runs impellerc manually, but ideally we'd be able to do that during runtime, so that the user can download any map style, not just prebaked ones.

Not sure if `FragmentProgram` and `flutter_gpu` use different API under the hood though. But considering the fact that Impeller is now stable on Android, this would be great to have.

**Comment by josxha** [comment]
Hello @jezell, could you please keep this feature request open? Even if the Flutter team has given it a low priority, it shows that this is a wanted feature by the community and let other users upvote.
